src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Runtime.ExceptionServices;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    class CallSiteRuntimeResolver : CallSiteVisitor<ServiceProvider, object>
+    {
+        public object Resolve(IServiceCallSite callSite, ServiceProvider provider)
+        {
+            return VisitCallSite(callSite, provider);
+        }
+
+        protected override object VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
+        {
+            return provider.CaptureDisposable(
+                VisitCallSite(transientCallSite.Service, provider));
+        }
+
+        protected override object VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
+        {
+            object[] parameterValues = new object[constructorCallSite.ParameterCallSites.Length];
+            for (var index = 0; index < parameterValues.Length; index++)
+            {
+                parameterValues[index] = VisitCallSite(constructorCallSite.ParameterCallSites[index], provider);
+            }
+
+            try
+            {
+                return constructorCallSite.ConstructorInfo.Invoke(parameterValues);
+            }
+            catch (Exception ex) when (ex.InnerException != null)
+            {
+                ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
+                // The above line will always throw, but the compiler requires we throw explicitly.
+                throw;
+            }
+        }
+
+        protected override object VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
+        {
+            return VisitScoped(singletonCallSite, provider.Root);
+        }
+
+        protected override object VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
+        {
+            object resolved;
+            lock (provider.ResolvedServices)
```

Comment:
This would result in two CAS checked property access instead one: one for `SyncRoot` other for `ResolvedServices`, we can remove `SyncRoot` and lock onto `ResolvedServices and `_transientDisposables` individually in service provider.